app/models/assignment.rb

```diff
@@ -457,7 +457,87 @@ def add_csv_group(row)
     collision_error
   end
+  def add_csv_groups(file)
+    group_list = []
+    group_search = []
+    grouping_list = []
+    grouping_search = []
+    student_list = []
+    CSV.foreach(file) do |row|
+      unless row.length.zero?
+        row.map! { |item| item.strip }
+      end 
+    # Since repo_name of "group" will be set before the first save call, the
+    # set repo_name will be used instead of the autogenerated name. See
+    # set_repo_name and build_repository in the groups model. Also, see
+    # create_group_for_working_alone_student in the students model for
+    # similar semantics.
+      if is_candidate_for_setting_custom_repo_name?(row)
+        # Do this only if user_name exists and is a student.
+        if Student.where(user_name: row[2]).first
+          repo_name = row[0]
+        else
+          # Student name does not exist, use provided repo_name
+          repo_name = row[1]
+        end
+      end
 
+    # If we are not repository admin, set the repository name as provided
+    # in the csv upload file
+      unless group.repository_admin?
+        repo_name = row[1]
+      end
+    # Note: We cannot use find_or_create_by here, because it has its own
+    # save semantics. We need to set and save attributes in a very particular
+    # order, so that everything works the way we want it to.
+      group = Group.where(group_name: row.first).first
+      if group.nil?
+        group_list << Group.new(group_name: row[0], repo_name: repo_name)
+        group_search << row[0]
+      end
+    # Note: after_create hook build_repository might raise
+    # Repository::RepositoryCollision. If it does, it adds the colliding
+    # repo_name to errors.on_base. This is how we can detect repo
+    # collisions here. Unfortunately, we can't raise an exception
+    # here, because we still want the grouping to be created. This really
+    # shouldn't happen anyway, because the lookup earlier should prevent
+    # repo collisions e.g. when uploading the same CSV file twice.
+    # unless group.errors[:base].blank?
+    #   collision_error = I18n.t('csv.repo_collision_warning',
+    #                       { repo_name: group.errors.on_base,
+    #                         group_name: row[0] })
+    # end
+    # Create a new Grouping for this assignment and the newly
+    # crafted group
+    # Form groups
+      start_index_group_members = 2 # first field is the group-name, second the repo name, so start at field 3
+      student_list << Student.where(user_name: row[start_index_group_members..(row.length - 1)])
+    end
+    Group.import group_list
+    groups = Group.where(group_name: group_search)
+    groups.each do |group|
+      group.set_repo_name
+      group.build_repository
+      grouping_list << Grouping.new(assignment: self, group: group, group_id: group.id)
+      grouping_search << group.id
+    end
+    Grouping.import grouping_list
+    groupings = Grouping.where(assignment_id: self.id, group_id: grouping_search) 
```

Comment:
Line is too long. [82/80]<br>Redundant `self` detected.<br>Trailing whitespace detected.